controller/src/test/java/io/pravega/controller/store/stream/HostStoreTest.java

```diff
@@ -89,20 +89,27 @@ public void zkHostStoreTests() {
                     .containerCount(containerCount)
                     .build();
 
+            // Update host store map.
+            Map<Host, Set<Integer>> hostContainerMap = HostMonitorConfigImpl.getHostContainerMap(host, controllerPort, containerCount);
+
             // Create ZK based host store.
             HostControllerStore hostStore = HostStoreFactory.createStore(hostMonitorConfig, storeClient);
-
-            CompletableFuture<Void> latch = new CompletableFuture<>();
+            CompletableFuture<Void> latch1 = new CompletableFuture<>();
+            CompletableFuture<Void> latch2 = new CompletableFuture<>();
             ((ZKHostStore) hostStore).addListener(() -> {
-                latch.complete(null);
+                // With the addition of updateMap() in tryInit(), this listener is actually called twice, and we need to
+                // wait for the second operation to complete (related to updateHostContainersMap()).
+                if (latch1.isDone()) {
+                    latch2.complete(null);
+                }
+                latch1.complete(null);
             });
-            // Update host store map.
-            Map<Host, Set<Integer>> hostContainerMap = HostMonitorConfigImpl.getHostContainerMap(host, controllerPort, containerCount);
             hostStore.updateHostContainersMap(hostContainerMap);
```

Comment:
A couple of clarification questions:
1- Does `updateHostContainersMap` trigger `tryInit`, and if not, which call actually triggers `tryInit`?
2- I understand that the addition of a second latch is dealing with with the second invocation of the listener, but the overall behavior looks strange, why is it that we need to invoke the listener twice in this scenario? Changing the underlying behavior is beyond the scope of this PR, but I'd rather not change a test case to comply with a bad behavior, so help me understand why we have this behavior in the first place.

Reply:
Let me address your questions:
1.- Yes, the invocation to `updateHostContainersMap` also includes a call to `tryInit()`, which now has a call to `updateMap()` that triggers the listener.
2.- Regarding the behavior that decides when to call the listener, it has been modified in PR #3682 by adding the call to `updateMap()` in `tryInit()`. That change induces an additional listener invocation (only the first time, when `zkInit=false`) to the one that is expected from the `NodeCache` object `hostContainerMapNode`. In this PR, we only deal with that behavior change in the test. If you prefer, we can modifying how listeners are invoked in `ZKHostStore`, but as far as I understand these listeners are only used for testing purposes (i.e., the behavior of the `ZKHostStore` should not be impacted when running Pravega). Please, let me know your preferred option.

Reply:
The listener is added in the code only for testing purposes.. 
and every time an update to the map happens, the listener notifies whoever is listening after the update..

The test makes an explicit update and expects a notification..

With new change of calling updatemap from tryInit, the there will be two notifications.. so the test should also expect two updates and hence two futures that should be completed